lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1254,6 +1254,36 @@ void OpenCLFunction::execute(ExecutionContext *context) {
       continue;
     }
 
+    if (auto *SLWSG = dyn_cast<SparseLengthsWeightedSumGradInst>(&I)) {
+      cl_kernel kernel = createKernel(kernelName);
+      // Set the device buffer as the first argument.
+      setKernelArg(kernel, 0, deviceBuffer_);
+      // Set all buffer arguments from the instruction (dataGrad, destGrad,
+      // weights, indices, lengths) as subsequent arguments.
+      auto numArgs = setKernelArgsForBuffers(kernel, I, 1, runtimeBundle_);
+
+      // Set the number of segments as the second last argument.
+      auto *lengths = SLWSG->getLengths();
+      size_t segments = lengths->size();
+      setKernelArg<cl_uint>(kernel, numArgs + 1, segments);
+
+      // Set the size of one slice of destGrad as the last argument.
+      auto *destGrad = SLWSG->getDestGrad();
+      size_t destGradSliceSize = destGrad->size() / destGrad->dims()[0];
+      setKernelArg<cl_uint>(kernel, numArgs + 2, destGradSliceSize);
+
+      // Zero the data gradient buffer so that the kernel can accumulate (+=)
+      // into it.
+      auto *dataGrad = SLWSG->getDataGrad();
+      fillBuffer(deviceBuffer_, runtimeBundle_.getValueOffset(dataGrad),
+                 dataGrad->size(), 0, dataGrad->getElementType());
+
+      // Enqueue the kernel.
+      enqueueKernel(I.getName(), commands_, kernel, deviceId_, {1},
```

Comment:
This `{1}` deserves a comment.

And BTW OpenCL supports atomics, so you could try to update dataGrad from multiple threads and see if it works and if improves the performance.